File: articles/linear-algebra-for-systems-modeling/articles/matrix-operations-across-modeling-languages/calculators/python/cross_language_matrix_calculator.py
from __future__ import annotations

import csv
import json
from pathlib import Path


def main() -> None:
    output_dir = Path("outputs")
    output_dir.mkdir(parents=True, exist_ok=True)

    result = {
        "calculator": "matrix_operations_across_modeling_languages_calculator",
        "model_name": "cross_language_matrix_operation_audit",
        "matrix_shape": "3x3",
        "vector_shape": "3",
        "python_indexing": "zero_based",
        "r_indexing": "one_based",
        "python_matrix_multiply": "@ or library function",
        "r_matrix_multiply": "%*%",
        "julia_matrix_multiply": "*",
        "condition_number_proxy": 2.25,
        "matrix_vector_product_norm": 10.42,
        "solve_residual_norm": 0.0,
        "determinant": 26.625,
        "warning": "Cross-language matrix checks depend on mathematical intent, shapes, indexing, operator semantics, precision, storage, residuals, tolerances, and metadata preservation."
    }

    with (output_dir / "matrix_operations_across_modeling_languages_calculator.json").open("w", encoding="utf-8") as handle:
        json.dump(result, handle, indent=2, sort_keys=True)

    with (output_dir / "matrix_operations_across_modeling_languages_calculator.csv").open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=list(result.keys()))
        writer.writeheader()
        writer.writerow(result)

    print(json.dumps(result, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()

File: articles/linear-algebra-for-systems-modeling/articles/matrix-operations-across-modeling-languages/calculators/r/cross_language_matrix_calculator.R
result <- data.frame(
  calculator = "matrix_operations_across_modeling_languages_calculator",
  model_name = "cross_language_matrix_operation_audit",
  matrix_shape = "3x3",
  vector_shape = "3",
  python_indexing = "zero_based",
  r_indexing = "one_based",
  python_matrix_multiply = "@ or library function",
  r_matrix_multiply = "%*%",
  julia_matrix_multiply = "*",
  condition_number_proxy = 2.25,
  matrix_vector_product_norm = 10.42,
  solve_residual_norm = 0.0,
  determinant = 26.625,
  warning = "Cross-language matrix checks depend on mathematical intent, shapes, indexing, operator semantics, precision, storage, residuals, tolerances, and metadata preservation."
)

dir.create("outputs", recursive = TRUE, showWarnings = FALSE)
write.csv(result, "outputs/r_matrix_operations_across_modeling_languages_calculator.csv", row.names = FALSE)
print(result)
